### PR TITLE
 add CRUD book

### PR DESCRIPTION
**Pull request by hald01**
add CRUD book

### CODE REVIEW

File: app/models/book.rb [hunk 1]
```diff
@@ -0,0 +1,7 @@
+class Book < ApplicationRecord
+  before_save { name.downcase! }
+
+  validates :name, presence: true, length: { maximum: 50 }, uniqueness: { case_sensitive: false }
+  validates :price, presence: true, length: { maximum: 20 }, numericality: { only_integer: true }
+  validates :number, presence: true, length: { maximum: 20 }, numericality: { only_integer: true }
```
Comment (vietle-myfave) [inline app/models/book.rb:6]: it's mean we can input the number with length is 20 when type is integer, right?

File: app/views/books/_book.html.erb [hunk 2]
```diff
@@ -0,0 +1,22 @@
+<p><pre>STT <%= book.id %></pre>
```
Comment (vietle-myfave) [inline app/views/books/_book.html.erb:1]: I think we should use English instead VietNamese :)))

File: app/views/books/_book.html.erb [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+      <br>
+      <label for="pwd">Number:</label>
+      <input type="integer" class="form-control" id="pwd">
+    </div>
+
+    <% if logged_in? && current_user.admin? %>
+      <br>
+      <%= link_to "Edit Book", edit_book_path(id: book.id, test: "test") %>
+      <br>
+      <%= link_to "Delete Book", book, data: { "turbo-method": :delete} %>
```
Comment (vietle-myfave) [inline app/views/books/_book.html.erb:18]: i think if we use bootstrap, we can you "mt-, mb-..." class to margin css instead use <br> tag.
What do you think?

File: app/controllers/books_controller.rb [hunk 4]
```diff
@@ -0,0 +1,51 @@
+class BooksController < ApplicationController
+  before_action :load_book, except: [:new, :create, :index]
```
Comment (vietle-myfave) [inline app/controllers/books_controller.rb:2]: i think we need to check authorization here.what do you think?

File: app/controllers/books_controller.rb [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+  def new
+    @book = Book.new
+  end
+
+  def create
+    @book = Book.new(book_params)
+    if @book.save
+      log_in @book
```
[the remark this answers is not in the document]
Comment (hald01) [reply app/controllers/books_controller.rb:11]: that is redundant, i will fix it

File: app/views/books/_book.html.erb [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+    <div class="form-group">
+      <button type="button" class="btn btn-info">Add to Card</button>
+      <br>
+      <label for="pwd">Number:</label>
+      <input type="integer" class="form-control" id="pwd">
+    </div>
+
+    <% if logged_in? && current_user.admin? %>
+      <br>
+      <%= link_to "Edit Book", edit_book_path(id: book.id, test: "test") %>
```
[the remark this answers is not in the document]
Comment (hald01) [reply app/views/books/_book.html.erb:16]: i will fix it

File: app/views/books/_form_edit.html.erb [hunk 7]
```diff
@@ -0,0 +1,14 @@
+<%= form_for(@book) do |f| %>
```
[the remark this answers is not in the document]
Comment (hald01) [reply app/views/books/_form_edit.html.erb:1]: i will fix it

File: app/views/books/index.html.erb [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+
+<% if logged_in? && current_user.admin? %>
+  <li><%= link_to "Add new Book", new_book_path %></li>
+<% end %>
+
+<%= will_paginate %>
+
+<%= render @books %>
+
+<%= will_paginate%>
```
[the remark this answers is not in the document]
Comment (hald01) [reply app/views/books/index.html.erb:12]: i want to have pagination top and bottom

File: app/views/books/new.html.erb [hunk 9]
```diff
@@ -0,0 +1,10 @@
+<% provide(:title, 'Add Book') %>
+<% provide(:button_text, 'Sumbit book') %>
```
[the remark this answers is not in the document]
Comment (hald01) [reply app/views/books/new.html.erb:2]: that rights, i will fix it

File: app/controllers/application_controller.rb [hunk 10]
```diff
@@ -1,3 +1,41 @@
...
+  end
+
+  def admin_user
+    return if current_user.admin?
+    flash[:success] = "only admin can see user list"
+    redirect_to error_path
+  end
+
+  def load_user    
+    user = User.find_by(id: params[:id])
```
[the remark this answers is not in the document]
Comment (hald01) [reply app/controllers/application_controller.rb:23]: i will fix it

File: app/controllers/application_controller.rb [hunk 11]
```diff
@@ -1,3 +1,41 @@
...
+    if user.present?
+      @user = user
+    else
+      flash[:danger] = "User id does not exits"
+      redirect_to error_path
+    end
+  end
+
+  def load_book    
+    book = Book.find_by(id: params[:id])
```
[the remark this answers is not in the document]
Comment (hald01) [reply app/controllers/application_controller.rb:33]: i will fix  it

File: app/models/book.rb (hunk 1 above)
Comment (hald01) [reply app/models/book.rb:6]: that  my mean, but i will fix it

File: app/views/books/_book.html.erb (hunk 2 above)
Comment (hald01) [reply app/views/books/_book.html.erb:1]: i will fix it

File: app/controllers/books_controller.rb (hunk 4 above)
Comment (hald01) [reply app/controllers/books_controller.rb:2]: i will fix it

File: app/views/books/_book.html.erb (hunk 3 above)
Comment (hald01) [reply app/views/books/_book.html.erb:18]: i will fix it. and i will add css after i complete all of function

File: app/controllers/users_controller.rb [hunk 12]
```diff
@@ -21,7 +21,7 @@ def create
     if @user.save
       log_in @user
       flash[:success] = "Welcome to the Book Store!"
-      redirect_to @user
+      redirect_to books_path
```
Comment (vietle-myfave) [inline app/controllers/users_controller.rb:24]: just wondering why we redirect to books screen after create a user

File: app/controllers/users_controller.rb [hunk 13]
```diff
@@ -21,7 +21,7 @@ def create
     if @user.save
       log_in @user
```
Comment (vietle-myfave) [inline app/controllers/users_controller.rb:22]: Why we have call login after create a user?

File: app/models/book.rb [hunk 14]
```diff
@@ -0,0 +1,7 @@
+class Book < ApplicationRecord
+  before_save { name.downcase! }
+
+  validates :name, presence: true, length: { maximum: 50 }, uniqueness: { case_sensitive: false }
+  validates :price, presence: true, length: { maximum: 10 }, numericality: { only_integer: true }
```
Comment (vietle-myfave) [inline app/models/book.rb:5]: the same for issue regarding out of range of integer with limit 4. I think it'll raise exception when user input a number with length is 10(example: 9999999999).you can try

File: app/views/books/_form.html.erb [hunk 15]
```diff
@@ -0,0 +1,14 @@
...
+  <%= f.label :name %>
+  <%= f.text_field :name, class: 'form-control' %>
+
+  <%= f.label :price %>
+  <%= f.text_field :price, class: 'form-control' %>
+
+  <%= f.label :number %>
+  <%= f.text_field :number, class: 'form-control' %>
+
+  <%= f.submit yield(:button_text), class: "btn btn-primary" %>
```
Comment (vietle-myfave) [inline app/views/books/_form.html.erb:13]: just wondering what the default name of button_text?

File: app/views/books/_form.html.erb [hunk 16]
```diff
@@ -0,0 +1,14 @@
+<%= form_for(@book) do |f| %>
```
Comment (vietle-myfave) [inline app/views/books/_form.html.erb:1]: are you sure that everything will be ok for both Update and Create?

File: db/migrate/20220413020752_add_index_to_books_name.rb [hunk 17]
```diff
@@ -0,0 +1,4 @@
+class AddIndexToBooksName < ActiveRecord::Migration[7.0]
```
[the remark this answers is not in the document]
Comment (vietle-myfave) [reply db/migrate/20220413020752_add_index_to_books_name.rb:1]: I think if nothing here, we no need this file

[the remark this answers is not in the document]
Comment (hald01) [reply db/migrate/20220413020752_add_index_to_books_name.rb:1]: i will fix it

**Pull request merged**